client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -185,10 +193,35 @@ private Checkpoint completeCheckpoint(String checkpointName) {
 
     @Override
     public void resetReaderGroup(ReaderGroupConfig config) {
+        manageSubscriptions(config);
```

Comment:
Here and where the group is initialized there is a risk of the application dying between updating one and the other.

One way to mitigate this without having the controller try to read the state is to make it eventually consistent by having the readers all attempt to update the controller when the notice the state has changed. This will result in multiple notifications but we can be confident the correct result will be reached.

Another approach might be to do the diff on the controller instead of on the client. Meaning that instead of sending subscribe and unsubscribe commands, the controller could simply be told: "This is the current config for x reader group" and then it would update all the subscriptions to the correct values. This is idempotent so it would be easy to simply retry if there was any problem.

Reply:
From what I understand, one of the approaches to this is to make this entire call using `synchroniser.updateState(...)` so that every reader can make an attempt to update the controller.

Reply:
Yes need to figure out the approach for handling partial failures.

Reply:
The problem is we need atomicity in two updates to two different entities (RG state and controller). And we need guarantee that until we have completed both steps, no new changes vis a vis subscription of this RG should be allowed in either of the entities. 

the first suggestion from tom has this problem that if all readers start reporting, some reader may report an older stale value which could overwrite a newer update in controller. 
so we will need to incorporate some notion of conditional update to controller for subscription. that is perhaps unnecessary and a departure from the way apis have been designed. 
perhaps a better approach would be to do the update to RG state in two phased protocol.. record the fact in RG state that we want to change its subscription.. then report that change to controller.. then update the RG state that the change has been reported successfully.. 
this means we have three steps.. and we want atomicity and failover. 

i had suggested earlier that we could perhpas elect a "leader" among the "readers" in the RG and designate it for such functions. It could be responsible for reporting both subscription changes and checkpoints. 
 
Since readers use statesynchronizer for coordinating. they could as well elect a leader from among them (for example - first one to update a specific field with its readerid in rg). 
we dont necessarily need a heartbeat mechanism for this leader. the reader can be assumed to be alive until it is explicitly shutdown or reported offline.. which is when we also reset the "leader" field to empty whereby triggering another election. 
To ensure that a reader which becomes the leader is always participating in consuming from the RG, we can also put an additional constraint that the reader participating in election should have at least one segment assigned to it. this takes care of a corner case where a reader with zero segment assignment crashed and no one bothered to report it offline and consumption was never affected. 

rough flow:
```
1. createRg state with leader = empty, work => subscribe
2. one or more addReader added to RG. each sees leader=empty and sees segments that it can acquire (unassigned segments) or already has segment assginment, then it attempts to update RG with leader="readerid" assignment=acquiredsegments etc.
3. leader reader's update succeeds and it then goes on to do the work --> subscribe/unsubscribe to streams.
4. resets RG state.work = empty
5. reports RG streamcut after each checkpoint completion. first updates work = report checkpoint. then reports checkpoint.. then resets work = empty. 
6. leader reader is reported offline ==> RG state.leader=empty
all readers that see RG leader=empty attempt to become a leader. 
```

Reply:
so far we have two types of works this leader would do:
```
enum Work {
Bootstrap,
Checkpoint
}
```

1. bootstrap work: subscribe/unsubscribe on streams
2. checkpoint work: report checkpoints for streams. 

Bootstrap of RG will require performing all functions like talking to controller to report subscription changes. this is posted when RG config created or changed (reset). When this work is performed, the input for this work is the current RG config. 
Question: do we allow RG state to be modified if an outstanding bootstrap work has not completed. IMO either is fine. 

Order of requests: do we care for order of bootstrap and checkpoint requests and want to capture it in RG state? perhaps yes. with just checkpoint and bootstrap we can choose not to worry about order. 
But since we are building this leader reader framework, in future it could be used for other types of work that a leader reader may have to partake. So we should keep an order of requests. Perhaps while ensuring that there is at max one outstanding request of each type of work. 
thoughts?

Reply:
@shiveshr we are discussing 2 separate  issues here:
1. All Controller update actions should be atomic - When we add/delete the Subscriber from multiple Streams, one of the API calls may fail, creating the need to reverse the successful calls (or rely on eventual consistency)

This can be solved by adding 3 new APIs to Controller:
a. addSubscriberToStreams(List<Stream>, String subscriber)
This internally uses the ControllerEventProcessor framework to make sure subscriber is added to all Streams (at least eventually)
b. deleteSubscriberFromStreams(List<Stream>, String subscriber)
Same like (a)
c. UpdateSubscribersForStreams(Map<Stream, String, opr>)
Add/delete subscriber from a stream and complete all updates atomically.

2. All Controller and State Synchronizer updates should be atomic
This can be solved by simply wrapping this code in a try/catch block and reversing the Controller updates in case of an exception.

Reply:
> All Controller update actions should be atomic 
thats a given.. i am not referring to this. 

> All Controller and State Synchronizer updates should be atomic
this is what i am talking about. 
try catch block doesnt solve it because your application can crash mid way during the "roll back" whereby leaving the state in partially completed state. which is not something that we want.

Reply:
Rather than attempting to re-invent two phase commit to try to achieve consistency between two different sources, how about we jut modify the controller API so that it is idempotent. 
The api "add x to y" and "remove x from z" are not idempotent. But if we just make the api: "Make x subscribed to y (and only y)" that is idempotent and we can just retry the request until it goes through.

Reply:
I have suggested modifying Controller API to make updates to all Streams ( for a Subscriber) idempotent. See: https://github.com/pravega/pravega/pull/5280/files#r518517888

In addition to this, on Client we'll be maintaining additional variables in RGState, to ensure failures during RGState reset/create do not cause the RGState to become inconsistent. Here is pseudocode for the same:
```
create
if (rg.state == null)
create rg state with isReady=false, toSubscribe=list<stream>
call controller to subscribe
update rg state with isReady=true
else
  do nothing.. idempotent create {
if (state.isReady=false) {
subscribe(state.toSubscribe)
update rg state with isReady=true, toSubscribe=null
}
reset(newResetConfig)
state = get RG state
 2. if (state.isReady = false) {
  subscribe(state.tosubscribe)
  unsubscribe(state.toUnsubscribe)
 updateconditionally(isReady=true, tosubscribe=null, tounsubscribe=null)
}
 3. if (state.config != newResetConfig) {
updateconditionally(newResetConfig, Ready=false, toSubcribe, toUnsubscribe, state.version)
  subscribe(state.tosubscribe)
  unsubscribe(state.toUnsubscribe)
}
```